src/daft-scan/src/expr_rewriter.rs

```diff
@@ -49,12 +49,84 @@ fn apply_partitioning_expr(expr: Expr, pfield: &PartitionField) -> Option<Expr>
     }
 }
 
+/// Grouping of clauses in a conjunctive predicate around partitioning semantics.
+pub struct PredicateGroups {
+    pub partition_filter: Vec<Expr>,
+    pub non_identity_partition_filter_for_data: Vec<Expr>,
+    pub partition_and_data_filter: Vec<Expr>,
+    pub data_only_filter: Vec<Expr>,
+}
+
+impl PredicateGroups {
+    pub fn new(
+        partition_filter: Vec<Expr>,
+        non_identity_partition_filter_for_data: Vec<Expr>,
+        partition_and_data_filter: Vec<Expr>,
+        data_only_filter: Vec<Expr>,
+    ) -> Self {
+        Self {
+            partition_filter,
+            non_identity_partition_filter_for_data,
+            partition_and_data_filter,
+            data_only_filter,
+        }
+    }
+
+    pub fn from_data_only(data_only_filter: Vec<Expr>) -> Self {
+        Self::new(vec![], vec![], vec![], data_only_filter)
+    }
+}
+
 pub fn rewrite_predicate_for_partitioning(
     predicate: Expr,
     pfields: &[PartitionField],
-) -> DaftResult<Option<Expr>> {
+) -> DaftResult<PredicateGroups> {
     if pfields.is_empty() {
-        return Ok(None);
+        return Ok(PredicateGroups::from_data_only(vec![predicate]));
+    }
+
+    let pfields_map: HashMap<&str, &PartitionField> = pfields
+        .iter()
+        .map(|pfield| (pfield.field.name.as_str(), pfield))
+        .collect();
+
+    // Before rewriting predicate for partition filter pushdown, partition predicate clauses into groups that will need
+    // to be applied at the data level (i.e. any clauses that aren't pure partition predicates with identity
+    // transformations).
+    let data_split = split_conjuction(&predicate);
+    // Predicates that only reference partition columns, but involve non-identity transformations.
+    let mut non_identity_part_preds_for_data = vec![];
+    // Predicates that reference both partition columns and data columns.
+    let mut part_and_data_preds = vec![];
+    // Predicates that only reference data columns (no partition column references).
+    let mut data_preds = vec![];
+    for e in data_split.into_iter() {
+        let mut all_data_keys = true;
+        let mut all_part_keys = true;
+        let mut any_non_identity_part_keys = false;
+        e.apply(&mut |e| {
+            if let Expr::Column(col_name) = e {
+                if let Some(pfield) = pfields_map.get(col_name.as_ref()) {
+                    all_data_keys = false;
+                    if !matches!(pfield.transform, Some(PartitionTransform::Identity) | None) {
```

Comment:
as discussed in prior comment, this will allow all predicate expressions though rather than only the ones through allowed operations